### ISSUE TITLE
Angular Rendering

### ISSUE THREAD

**Issue by stevenfollis** [johntitus/node-horseman#167]
Hello! Love how clear node-horseman is to use compared to other frameworks, but am having an issue. I have an AngularJS 1.5.2 app with a view that needs to be exported to PDF.  

Right now, node-horseman returns the index.html page, but before angular renders and binds everything up.  

Is there something specific that I need to do to retrieve the fully rendered page??  Have tried many of the wait methods without much luck.  Thanks!

**Comment by johntitus** [comment]
After the open() is complete, could you waitForSelector() for something to
show up on the page?

**Comment by awlayton** [comment]
I do not know of a generic way for the browser to know when a page is "rendered". I've had these sorts of problems with screenshotting sites using React, Angular etc.

If you control the page, you could perhaps have it add a hidden DOM element when rendering is done and then have horseman wait for that element.

If you just care about it working, you could add a `.delay(/*some ms*/)` and just give the rendering time to happen.

**Comment by stevenfollis** [comment]
Have tried .waitForSelector() for an image that renders inside of an ng-view, but I kept getting timeouts. 

I think the issue may be related to the ng-route module that swaps out views dynamically? Have also tried .delay() with large numbers and still get the base html page without the rendered content.  I'll keep plugging away.

**Comment by awlayton** [comment]
If you can post a URL and your code I can take a look at it this weekend probably.

**Comment by awlayton** [comment]
Have you tried running it with `DEBUG='horseman*'`? It could be that an error is occur so the page just never renders.

**Comment by stevenfollis** [comment]
DEBUG is quite helpful! Thanks for that.  Looks I'm getting a rendering error:

`horseman:verbose onError Error: [$compile:tpload] http://errors.angularjs.org/1.5.2/$compile/tpload?p0=app%2Fhome%2Fhome.html&p1=undefined&p2=undefined`

This being somehow related to the template system in Angular.  Odd, the page works in every browser/configuration I've tried.  Curious why PhantomJS isn't a fan. 

Would send a URL but it's behind an authorization wall.

**Comment by awlayton** [comment]
Perhaps it is trying to use some functionality missing in PhantomJS. You could try polyfilling like I showed in #164.

**Comment by stevenfollis** [comment]
Doesn't seem to have made a difference, but was worth a try. Still erroring out.  Tried running vanilla PhantomJS from the command line and seeing:

```
Error: [$compile:tpload] Failed to load template: app/home/home.html (HTTP status: undefined undefined)
http://errors.angularjs.org/1.5.5/$compile/tpload?p0=app%2Fhome%2Fhome.html&p1=undefined&p2=undefined
handleError@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:18979:47
processQueue@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:15961:30
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:15977:39
$eval@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:17229:28
$digest@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:17045:36
$apply@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:17337:31
bootstrapApply@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:1749:21
invoke@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:4665:24
doBootstrap@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:1747:20
bootstrap@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:1767:23
angularInit@http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:1652:14
http://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js:30863:16
i@http://code.jquery.com/jquery-2.2.2.min.js:2:27420
fireWith@http://code.jquery.com/jquery-2.2.2.min.js:2:55176
ready@http://code.jquery.com/jquery-2.2.2.min.js:2:29982
J@http://code.jquery.com/jquery-2.2.2.min.js:2:30161 undefined
```

Some template isn't compiling, but i can't get insight into why.  Can't repro it in a browser manually.

**Comment by awlayton** [comment]
Have you tried disabling web security @stevenfollis? Some of my googling make it sounds like that could be involved.

**Comment by stevenfollis** [comment]
Main app is deeply hooked into auth, not an easy way to disable. However, I did standup a skeleton POC and it does look to be an issue with [azure-activedirectory-library-for-js](https://github.com/AzureAD/azure-activedirectory-library-for-js), the technique I'm using to authenticate the Angular SPA.  I had tested the server side authentication components and it looked OK, but yep I think this is JS related. Thanks for the help!
